### PR TITLE
Totally rework Guice JPA library

### PR DESCRIPTION
**Pull request by xrcat**
-- Introduce @DB(transaction = REQUIRES_NEW) mode
-- Introduce @DB.retries to automatically restart transactions after dead locks
-- Introduce manual db-context handling using UnitOfWork.begin()/end()
-- Use injector-detached method interceptors which
     clears the way to compile-time instrumentation
-- Rework multi-unit persistence concept, now persistent-units are related
     to EntityManager, not the annotated method

### CODE REVIEW

**Comment by xrcat** [comment]
@akochurov ?

File: guice-jpa/pom.xml [hunk 1]
```diff
@@ -18,10 +18,5 @@
             <artifactId>hibernate-jpa-2.0-api</artifactId>
             <scope>provided</scope>
         </dependency>
-        <!-- TODO: get rid of cglib-->
```
Comment (akochurov) [inline guice-jpa/pom.xml]: Cool!

File: guice-jpa/src/main/java/com/maxifier/guice/jpa/DB.java [hunk 2]
```diff
@@ -4,26 +4,48 @@
...
+     * Rollbacks transaction if method invocation ended with exception and commits otherwise.</li>
+     * </ul>
+     */
     Transaction transaction() default Transaction.NOT_REQUIRED;
 
-    public enum Transaction {
+    /**
+     * Automatically retries marked method in case of {@link java.sql.SQLTransientException}.
+     * <p>Use with extreme caution. Method retry may cause side effects with unpredictable results.</p>
+     * <p>This functionality designed to cope with db deadlocks which unavoidable in production environments.
```
Comment (akochurov) [inline guice-jpa/src/main/java/com/maxifier/guice/jpa/DB.java]: which are

File: guice-jpa/src/main/java/com/maxifier/guice/jpa/DBEntityManagerProvider.java [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+
+    @Override
+    public EntityManager get() {
+        return proxy;
+    }
+
+    @Override
+    public Object invoke(Object proxy, Method method, Object[] args) throws Throwable {
+        if (method.getDeclaringClass() == Object.class) {
+            return method.invoke(this, args); // Don't proxy Object's methods
```
Comment (akochurov) [inline guice-jpa/src/main/java/com/maxifier/guice/jpa/DBEntityManagerProvider.java]: I guess you should delegate toString too, no?

Comment (akochurov) [inline guice-jpa/src/main/java/com/maxifier/guice/jpa/DBEntityManagerProvider.java]: I mean
```java
if (method.equals(TO_STRING)) {
  return this.toString();
}
```

File: guice-jpa/src/main/java/com/maxifier/guice/jpa/DBInterceptor.java [hunk 4]
```diff
@@ -1,342 +1,98 @@
...
-        });
-        wrapper = (EntityManager) enhancer.create();
-        prepared = true;
-    }
-
+public class DBInterceptor implements MethodInterceptor {
+    private static final Logger logger = LoggerFactory.getLogger(DBInterceptor.class);
+    private static final Random RND = new Random();
+    private static final int RETRY_TIMEOUTS[] = new int[]{
+        0, 307, 2000, 11000, 19000, 31000, 53000, 89000, 151000, 241000, 307000
```
Comment (akochurov) [inline guice-jpa/src/main/java/com/maxifier/guice/jpa/DBInterceptor.java]: ExponentialBackOff?

File: guice-jpa/src/main/java/com/maxifier/guice/jpa/UnitOfWork.java [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+    }
+
+    @Nullable
+    static UnitOfWork get() {
+        return current.get();
+    }
+
+    UnitOfWork() {
+        this.previous = current.get();
+        current.set(this);
```
Comment (akochurov) [inline guice-jpa/src/main/java/com/maxifier/guice/jpa/UnitOfWork.java]: This seems to be hacky and too implicit. Should we instead move this out of constructor?

File: guice-jpa/src/test/java/com/maxifier/guice/jpa/DBEntityManagerProviderTest.java [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+import javax.persistence.EntityManager;
+import javax.persistence.EntityManagerFactory;
+import javax.persistence.metamodel.Metamodel;
+
+import static org.mockito.Mockito.*;
+
+/**
+ * @author Konstantin Lyamshin (2015-11-17 0:15)
+ */
+public class DBEntityManagerProviderTest extends org.testng.Assert {
```
Comment (akochurov) [inline guice-jpa/src/test/java/com/maxifier/guice/jpa/DBEntityManagerProviderTest.java]: God, extends Assert??? Use static import!

**Comment by akochurov** [comment]
LGTM. However there's a minor issue with toString in EntityManager proxy!

File: guice-jpa/src/main/java/com/maxifier/guice/jpa/DB.java (hunk 2 above)
Comment (xrcat) [inline guice-jpa/src/main/java/com/maxifier/guice/jpa/DB.java]: Don't understand sorry.

File: guice-jpa/src/main/java/com/maxifier/guice/jpa/DBEntityManagerProvider.java (hunk 3 above)
Comment (xrcat) [inline guice-jpa/src/main/java/com/maxifier/guice/jpa/DBEntityManagerProvider.java]: toString.getDeclarigClass() == Object. toString don't passed to actual EntityManager.

File: guice-jpa/src/main/java/com/maxifier/guice/jpa/DBInterceptor.java (hunk 4 above)
Comment (xrcat) [inline guice-jpa/src/main/java/com/maxifier/guice/jpa/DBInterceptor.java]: Yes this is reinvented exponential back off. Are there some standard implementations?

File: guice-jpa/src/main/java/com/maxifier/guice/jpa/UnitOfWork.java (hunk 5 above)
Comment (xrcat) [inline guice-jpa/src/main/java/com/maxifier/guice/jpa/UnitOfWork.java]: ok

File: guice-jpa/src/test/java/com/maxifier/guice/jpa/DBEntityManagerProviderTest.java (hunk 6 above)
Comment (xrcat) [inline guice-jpa/src/test/java/com/maxifier/guice/jpa/DBEntityManagerProviderTest.java]: Why? Extend works perfect and takes aways problems with Idea auto import optimization.

File: guice-jpa/src/main/java/com/maxifier/guice/jpa/DBEntityManagerProvider.java (hunk 3 above)
Comment (akochurov) [inline guice-jpa/src/main/java/com/maxifier/guice/jpa/DBEntityManagerProvider.java]: Ok

File: guice-jpa/src/main/java/com/maxifier/guice/jpa/DB.java (hunk 2 above)
Comment (akochurov) [inline guice-jpa/src/main/java/com/maxifier/guice/jpa/DB.java]: 'are' is missing

File: guice-jpa/src/main/java/com/maxifier/guice/jpa/DBInterceptor.java (hunk 4 above)
Comment (akochurov) [inline guice-jpa/src/main/java/com/maxifier/guice/jpa/DBInterceptor.java]: Oops I thought there's one in Guava, but it is actually situated in DFP API lib...

File: guice-jpa/src/main/java/com/maxifier/guice/jpa/DB.java (hunk 2 above)
Comment (xrcat) [inline guice-jpa/src/main/java/com/maxifier/guice/jpa/DB.java]: :+1:

**Pull request merged**